spring-pulsar/src/test/java/org/springframework/pulsar/core/PulsarAdministrationTests.java

```diff
@@ -0,0 +1,116 @@
+package org.springframework.pulsar.core;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import java.util.List;
+import java.util.Map;
+import org.apache.pulsar.client.admin.PulsarAdmin;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.springframework.beans.factory.ObjectProvider;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration
+public class PulsarAdministrationTests extends AbstractContainerBaseTests {
+
+	private static final String NAMESPACE = "public/default";
+
+	@Autowired
+	private ApplicationContext applicationContext;
+
+	@Configuration(proxyBeanMethods = false)
+	static class AdminConfiguration {
+		@Bean
+		PulsarAdministration admin() {
+			return new PulsarAdministration(Map.of("serviceUrl", getHttpServiceUrl()));
+		}
+	}
+
+	List<String> getTopics() throws Exception {
+		try (PulsarAdmin admin = PulsarAdmin.builder().serviceHttpUrl(getHttpServiceUrl()).build()) {
+			return admin.topics().getList(NAMESPACE);
+		}
+	}
+
+	void assertTopicsExist(List<String> actualTopics, ObjectProvider<PulsarTopic> expected) {
+		List<String> expectedTopics= expected.stream()
+				.<String>mapMulti((topic, consumer) -> {
```

Comment:
Ah nice! Clever use of `mapMulti`. **TIL**.

What you have done w/ this test is an excellent application of the nested test stuff we were chatting about offline. I took the liberty to take what you have here and tweak it a bit to come w/ something that simplifies it a bit. I will post the code here in its entirety rather than try to comment line by line by line. 

```java
package org.springframework.pulsar.core;

import static org.assertj.core.api.Assertions.assertThat;
import java.util.List;
import java.util.Map;
import org.apache.pulsar.client.admin.PulsarAdmin;
import org.apache.pulsar.client.admin.PulsarAdminException;
import org.apache.pulsar.client.api.PulsarClientException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.ObjectProvider;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ApplicationContext;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit.jupiter.SpringExtension;


@ExtendWith(SpringExtension.class)
@ContextConfiguration
public class PulsarAdministrationTests extends AbstractContainerBaseTests {

	private static final String NAMESPACE = "public/default";

	@Autowired
	private PulsarAdmin pulsarAdminClient;

	@Configuration(proxyBeanMethods = false)
	static class AdminConfiguration {
		@Bean
		PulsarAdmin pulsarAdminClient() throws PulsarClientException {
			return PulsarAdmin.builder().serviceHttpUrl(getHttpServiceUrl()).build();
		}
	}

	private void assertThatTopicsExist(List<PulsarTopic> expected) throws PulsarAdminException {
		List<String> expectedTopics= expected.stream()
				.<String>mapMulti((topic, consumer) -> {
					if (topic.isPartitioned()) {
						for (int i = 0; i < topic.numberOfPartitions(); i++) {
							consumer.accept(topic.getFullyQualifiedTopicName() + "-partition-" + i);
						}
					} else {
						consumer.accept(topic.getFullyQualifiedTopicName());
					}

				}).toList();
		assertThat(pulsarAdminClient.topics().getList(NAMESPACE)).containsAll(expectedTopics);
	}

	@Nested
	@ContextConfiguration(classes = CreateMissingTopicsTest.CreateMissingTopicsConfig.class)
	class CreateMissingTopicsTest {

		@Test
		void testTopics(@Autowired ObjectProvider<PulsarTopic> expectedTopics) throws Exception {
			assertThatTopicsExist(expectedTopics.stream().toList());
		}

		@Configuration(proxyBeanMethods = false)
		static class CreateMissingTopicsConfig {

			@Bean
			PulsarTopic nonPartitionedTopic() {
				return PulsarTopic.builder("cmt-non-partitioned-1").build();
			}

			@Bean
			PulsarTopic nonPartitionedTopic2() {
				return PulsarTopic.builder("cmt-non-partitioned-2").build();
			}

			@Bean
			PulsarTopic partitionedTopic() {
				return PulsarTopic.builder("cmt-partitioned-1").setNumberOfPartitions(4).build();
			}
		}

	}

	@Nested
	@ContextConfiguration(classes = IncrementPartitionCountTest.IncrementPartitionCountConfig.class)
	class IncrementPartitionCountTest {

		@BeforeEach
		void createSmallerTopic() {
			try {
				pulsarAdminClient.topics().createPartitionedTopic("ipc-partitioned-1", 2);
			} catch (Exception e) {
				throw new RuntimeException(e);
			}
		}
		
		@Test
		void testTopics(@Autowired ObjectProvider<PulsarTopic> expectedTopics) throws Exception {
			assertThatTopicsExist(expectedTopics.stream().toList());
		}

		@Configuration(proxyBeanMethods = false)
		static class IncrementPartitionCountConfig {
			@Bean
			PulsarTopic partitionedTopic() {
				return PulsarTopic.builder("ipc-partitioned-1").setNumberOfPartitions(4).build();
			}
		}
	}
}

```

As you can see, by lifting up the PulsarAdmin into an actual bean and self-injecting at the top-level we can just use that in our assert method now. Also, the static was replaced w/ a BeforeEach. 

Let me know what you think.

Reply:
Thanks Chris, this definitely looks better than my static setup and makes the test easier to understand 👍